File: neet150.java
/**
 * https://neetcode.io/roadmap
 */

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class neet150 {

    public static void main(String[] args) {
        // int[] nums = {1, 2, 3, 3};
        // System.out.println(hasDuplicate(nums));
        // int[] nums1 = {1, 2, 3, 4};
        // System.out.println(hasDuplicate(nums1));

        // String s = "racecar";
        // String t = "carrace";
        // System.out.println(isAnagram(s, t));
        // s = "jar";
        // t = "jam";
        // System.out.println(isAnagram(s, t));
        // s = "a";
        // t = "ab";
        // System.out.println(isAnagram(s, t));
    }
    
    public static boolean hasDuplicate(int[] nums) {
        Set<Integer> numsSet = new HashSet<Integer>(); 
        for (int i = 0; i < nums.length; i++) {
            numsSet.add(nums[i]);
        }
        return numsSet.size() != nums.length;
    }

    public static boolean isAnagram(String s, String t) {
        if (s.length() != t.length()) {
            return false;
        }
        List<Character> s_letters = new ArrayList<Character>();
        for (char c : s.toCharArray()) {
            s_letters.add(c);
        }
        for (char charT : t.toCharArray()) {
            
            if (s_letters.contains(charT)) {
                s_letters.remove((Character) charT);
            }
        }
        return s_letters.size() == 0;
    }

    public int[] twoSum(int[] nums, int target) {
        int nums_len = nums.length;
        int count = 0;
        for (int i = nums_len - 1; i > 0; i--){
            for (int j = 0; j < i; j++) {
                if (nums[nums_len - i - 1] + nums[j + 1 + count] == target) {
                    int[] output = {nums_len - i - 1, j + 1 + count};
                    return output;
                }
            }
            count += 1;
        }
        int[] output = {};
        return output;
    }
}

File: neet150.py
# https://neetcode.io/roadmap

from typing import List
from collections import Counter

class Solution:
    def hasDuplicate(self, nums: List[int]) -> bool:
        counter = Counter(nums)
        
        return len(counter) != len(nums)
    
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False
        s_letters = list(s)
        for char in t:
            if char in s_letters:
                s_letters.remove(char)

        return len(s_letters) == 0
    
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        nums_len = len(nums)
        count = 0
        for i in range(nums_len - 1, 0, -1):
            for j in range(i):
                if nums[nums_len - i - 1] + nums[j + 1 + count] == target:
                    return[nums_len - i - 1, j + 1 + count]
            count += 1


def main():
    solution = Solution()
    # nums = [1, 2, 3, 3]
    # print(solution.hasDuplicate(nums))
    # nums = [1, 2, 3, 4]
    # print(solution.hasDuplicate(nums))
    s = "racecar"
    t = "carrace"
    print(solution.isAnagram(s, t))
    s = "jar"
    t = "jam"
    print(solution.isAnagram(s, t))
    s = "a"
    t = "ab"
    print(solution.isAnagram(s, t))

main()